### ISSUE TITLE
[META] CIBox 2.0

### ISSUE THREAD

**Issue by Sanchiz** [cibox/cibox#468]
Let's discuss possibility to implement CIBox architecture in new way.

We can decouple project and care only about reinstall script which we need for builds/deployments.

We have following parts:
## Docroot
- should support different CMF/Frameworks - Drupal, WordPress, Symfony and etc.
- should support multiple docroots

## VirtualMachine
- should support different virtual machines - CIBoxVM, DrupalVM, docker and etc.
- should know where it can find docroots and what types they are

## Builds
- CIBox should provide `install` scripts and `.cibox.yml` with configurations
- builds should be build on local environment or remote(docker style), e.g. environment independent
 
The main idea to maintain only `.cibox.yml` and `install` scripts and use docker for builds.

Here's diagram:
![image](https://cloud.githubusercontent.com/assets/1316234/15805280/e213c43c-2b2d-11e6-99f5-65f446dab025.png)

**Comment by Sanchiz** [comment]
@asgorobets please add your thoughts.

**Comment by asgorobets** [comment]
Agreed with above statements, after a talk with @Sanchiz at DrupalCamp Kiev we came up with some ideas I will promptly document here, but they all need refinement and separate tickets:

- We should start with refining the idea of .cibox.yml file and it's contents, as it's the part that user will interact with, and there are couple of good services that provide very nice examples. I like platform.sh approach, travis, shippable and others alike, I think this will be a big discussion and can take part in separate ticket

- CMF/Framework part of CIBOX should not be hardcoded/assumed by other parts of it, we need to have a way to assemble any project on any PHP framework just by specifying a "flavor" in .cibox.yml. Flavors can be drupal7, drupal8, wordpress, symfony, php and potentially extendable. Idea is to allow for custom flavors if someone can provide a plugin/implementation of it's own describing flavor specifics, but this can be done later on. There can be flavor specific configs, here comes profile and sql flow in Drupal, or just SQL flow in Wordpress.

- Local VirtualMachine or Docker
We need to start documenting requirements for the virtualmachine and see every variable paths/parts that different machines have and allow for those to be set from .cibox.yml if needed. Ideally local environment should be close to the builds environment and close to production, but doing that using Ansible provisioning was and is the **biggest** pain for the local VM part of the project, the provisioning for every developer becomes pain and is very error prone. We need to allow environment requirements to be carried by other VMs and potentially still support our own one, but maybe not enforce it and provide alternatives. if we do the CI builds using docker, same docker images can be deployed on local, for those of us who want to get a very close to builds environment.

- CI Build and Jenkins
Probably the most important part of CIBOX is it's Jenkins and Builds capabilities, builds take advantage of preinstalled software on the server and expose an environment for users to interact with available on a URL. The biggest concerns in past years to me was that it is not possible to change that environment in any way other than doing your own provision for CIBOX or start maintaining software on the server by yourself. There is no option for some specific version of PHP, nor for OS alternatives as some enterprise customers require. One way to solve this is to use Docker containers and Docker composer to describe an environment and deploy infrastructure based on some predefined docker composer files + ability to customize the environment by doing overrides per project, in case someone needs Solr or Elasticsearch, still the basic distro will be unified and supported by CIBOX team with ability to have a fixed version of docker image that can be deployed on other CIBOX servers without doing the provisioning again and getting unknown errors every time it's running. CIBOX maintainers will take care of building new images with useful defaults but still keep the update/upgrade path easy and maintainable.
A security update for CIBOX instance should be as easy as redeploying new docker image. We can adopt some docker-compose like format on top of docker-compose that will define our services we want, examples are platform.sh, terra-cli environment format, travis does the same:
```
os: centos:5.2
php: 5.6
flavor: drupal
drupal-version: 7
drush: 8.1.*

services:
- database: mysql:5.1
- elasticsearch: 2.3.3
```

I propose to start decomposing those tickets in separate discussions and start refining some tools/practices we're willing to adopt.

My favorite candidates for inspiration are:
- [Kalabox](http://support.kalabox.io/support/solutions)
- [Terra CLI](https://github.com/terra-ops/terra-cli)
- [Platform.sh](https://platform.sh/)

**Comment by Sanchiz** [comment]
Also some other links:
- https://app.shippable.com
- http://www.alpinelinux.org
